app/code/Magento/Quote/Model/CustomerManagement.php

```diff
@@ -100,11 +105,47 @@ public function populateCustomerInfo(QuoteEntity $quote)
         }
     }
 
+    /**
+     * Validate 'CustomerAddressId' in quote for a newly created customer.
+     *
+     * @param QuoteEntity $quote
+     * @return void
+     */
+    private function validateCustomerAddressId(QuoteEntity $quote): void
+    {
+        $customer = $quote->getCustomer();
+
+        if ($customer->getDefaultBilling()) {
+            $quote->getBillingAddress()->importCustomerAddressData(
+                $this->customerAddressRepository->getById($customer->getDefaultBilling())
+            );
+            $quote->getBillingAddress()->setCustomerAddressId($customer->getDefaultBilling());
+        } else {
+            $quote->getBillingAddress()->setCustomerAddressId(0);
+        }
+
+        if ($quote->getShippingAddress()->getSameAsBilling() && $customer->getDefaultBilling()) {
+            $quote->getShippingAddress()->importCustomerAddressData(
+                $this->customerAddressRepository->getById($customer->getDefaultBilling())
+            );
+            $quote->getShippingAddress()->setCustomerAddressId($customer->getDefaultBilling());
+        }
+
+        if (!$quote->getShippingAddress()->getSameAsBilling() && $customer->getDefaultShipping()) {
+            $quote->getShippingAddress()->importCustomerAddressData(
+                $this->customerAddressRepository->getById($customer->getDefaultShipping())
+            );
+            $quote->getShippingAddress()->setCustomerAddressId($customer->getDefaultShipping());
+        } else {
+            $quote->getShippingAddress()->setCustomerAddressId(0);
+        }
```

Comment:
I see quite similar code in https://github.com/magento/magento2/pull/29139/files#diff-9d928aadb5e8377f0a87b2ca364c48b7R91-R105
Is there any way to not copy paste, but unify it for all needed cases instead? I don't believe we need to run it twice.